### PR TITLE
Ahcorde/add/joint limit interface

### PR DESCRIPTION
**Pull request by guru-florida**
As suggested by @destogl, this is a follow-on to a PR by ahcorde, [PR135](https://github.com/ros-controls/ros2_control/pull/135).

This PR has joint_limit_interface updated with #134 changes which remote state and command handles and replace with just JointHandle. It has been synced with the latest ros2_control/master branch.

Also all tests (previously commented out) have been fixed up and pass on my local system.

Suggesting review by @Karsten1987 @bmagyar as well.

This PR:
Originally coded  by: @ddengster ed.fan@osrfoundation.org
Signed-off-by: @ahcorde ahcorde@gmail.com
#134, refactor + unit tests by myself: guru-florida

I believe ahcorde and I are both working to the same goal of getting gazeobo+ros+ros-control working together! :) Kudos to all, we are so close!

FYI I also have transmission_interface similarly updated with master in [PR182](https://github.com/ros-controls/ros2_control/pull/182). This PR is also included in PR182 so you can just approve and do that one, or both, PR182 is based from 181 and both synced to master so no worries. It's just a matter of scope.

### CODE REVIEW

**Comment by guru-florida** [comment]
LICENSING FYI @v-lopez @destogl @bmagyar   I changed the file header licenses to Apache 2 as [discussed here](https://github.com/ros-controls/ros2_control/pull/135#pullrequestreview-492192473).

File: joint_limits_interface/CHANGELOG.rst [hunk 1]
```diff
@@ -0,0 +1,247 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Changelog for package joint_limits_interface
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (AndyZe) [inline joint_limits_interface/CHANGELOG.rst:3]: IMO you could drop the ROS1 changelog, especially since the versions won't line up

Comment (guru-florida) [reply joint_limits_interface/CHANGELOG.rst]: Makes sense. done.

**Comment by bmagyar** [comment]
@ahcorde we'd appreciate your feedback on this one

**Review by ahcorde — APPROVED** [review]
LGTM,  @bmagyar should review the maintainers in the `package.xml`

**Comment by guru-florida** [comment]
@bmagyar @Karsten1987 @destogl In this refactor for #134 I went with JointHandle arguments wrapped in  std::shared_ptr<>, should these have been without shared_ptr? This was due to a commit I was using as a reference that had shared_ptr, but after reviewing more code such as robot_hardware.cpp I dont see any reason to use shared_ptr here, the JontHandle by value would work fine and probably more compatible. Thoughts?

File: joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+      max_pos_limit_ = std::numeric_limits<double>::max();
+    }
+  }
+
+/**
+ * \brief Enforce position and velocity limits for a joint that is not subject to soft limits.
+ *
+ * \param period Control period.
+ */
+  void enforceLimits(const rclcpp::Duration & period)
```
Comment (bmagyar) [inline joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp:201]: Refactor this name

File: joint_limits_interface/package.xml [hunk 3]
```diff
@@ -0,0 +1,39 @@
+<package format="3">
+  <name>joint_limits_interface</name>
+  <version>0.16.0</version>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:3]:
```suggestion
  <version>0.0.0</version>
```

File: joint_limits_interface/include/joint_limits_interface/joint_limits.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2013, PAL Robotics S.L.
```
Comment (bmagyar) [inline joint_limits_interface/include/joint_limits_interface/joint_limits.hpp:1]: we need to update all the licenses in joint_limits_interface

File: joint_limits_interface/package.xml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  <description>Interface for enforcing joint limits.</description>
+
+  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
+  <maintainer email="enrique.fernandez.perdomo@gmail.com">Enrique Fernandez</maintainer>
+  <maintainer email="mathias.luedtke@ipa.fraunhofer.de">Mathias Lüdtke</maintainer>
+
+  <license>BSD</license>
+
+  <url type="website">https://github.com/ros-controls/ros_control/wiki</url>
+  <url type="bugtracker">https://github.com/ros2-controls/ros_control/issues</url>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:13]:
```suggestion
  <url type="bugtracker">https://github.com/ros-controls/ros2_control/issues</url>
```

File: joint_limits_interface/package.xml [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  <url type="repository">https://github.com/ros2-controls/ros_control</url>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:14]:
```suggestion
  <url type="repository">https://github.com/ros-controls/ros2_control</url>
```

File: joint_limits_interface/package.xml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  <version>0.16.0</version>
+  <description>Interface for enforcing joint limits.</description>
+
+  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
+  <maintainer email="enrique.fernandez.perdomo@gmail.com">Enrique Fernandez</maintainer>
+  <maintainer email="mathias.luedtke@ipa.fraunhofer.de">Mathias Lüdtke</maintainer>
+
+  <license>BSD</license>
+
+  <url type="website">https://github.com/ros-controls/ros_control/wiki</url>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:12]:
```suggestion
  <url type="website">https://github.com/ros-controls/ros2_control/wiki</url>
```

File: joint_limits_interface/package.xml [hunk 8]
```diff
@@ -0,0 +1,39 @@
+<package format="3">
+  <name>joint_limits_interface</name>
+  <version>0.0.0</version>
+  <description>Interface for enforcing joint limits.</description>
+
+  <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
+  <maintainer email="enrique.fernandez.perdomo@gmail.com">Enrique Fernandez</maintainer>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:7]:
```suggestion
  <maintainer email="jordan.palacios@pal-robotics.com">Jordan Palacios</maintainer>
```

File: joint_limits_interface/package.xml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  <maintainer email="mathias.luedtke@ipa.fraunhofer.de">Mathias Lüdtke</maintainer>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:8]:
```suggestion
```

File: joint_limits_interface/package.xml [hunk 10]
```diff
@@ -0,0 +1,38 @@
+<package format="3">
+  <name>joint_limits_interface</name>
```
Comment (bmagyar) [inline joint_limits_interface/package.xml:2]: Need to add joint_limits_interface to linting jobs on CI

**Review by bmagyar — APPROVED** [review]

File: joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp (hunk 2 above)
Comment (bmagyar) [reply joint_limits_interface/include/joint_limits_interface/joint_limits_interface.hpp:201]: https://github.com/ros-controls/ros2_control/issues/188

File: joint_limits_interface/include/joint_limits_interface/joint_limits.hpp (hunk 4 above)
Comment (bmagyar) [reply joint_limits_interface/include/joint_limits_interface/joint_limits.hpp:1]: https://github.com/ros-controls/ros2_control/issues/189

File: joint_limits_interface/package.xml (hunk 10 above)
Comment (bmagyar) [reply joint_limits_interface/package.xml:2]: https://github.com/ros-controls/ros2_control/issues/190

File: joint_limits_interface/package.xml (hunk 7 above)
Comment (guru-florida) [reply joint_limits_interface/package.xml]: FYI There is no ros2_controls Wiki yet, it just goes to the main code/README page

**Pull request merged**